micro-rdk/src/common/provisioning/server.rs

```diff
@@ -0,0 +1,528 @@
+#![allow(dead_code)]
+use std::{marker::PhantomData, pin::Pin, rc::Rc, task::Poll};
+
+use bytes::{BufMut, Bytes, BytesMut};
+use futures_lite::{Future, FutureExt};
+use http_body_util::BodyExt;
+use hyper::{
+    body::Incoming, header::CONTENT_TYPE, http, rt, server::conn::http2, service::Service, Request,
+    Response,
+};
+use prost::Message;
+
+use crate::{
+    common::{
+        grpc::{GrpcBody, GrpcError, GrpcResponse},
+        webrtc::api::AtomicSync,
+    },
+    proto::provisioning::{
+        self,
+        v1::{
+            GetSmartMachineStatusResponse, SetSmartMachineCredentialsRequest,
+            SetSmartMachineCredentialsResponse,
+        },
+    },
+};
+
+use super::storage::Storage;
+
+struct ProvisioningServiceBuilder {
+    last_connection_attempt: Option<NetworkInfo>,
+    provisioning_info: Option<ProvisioningInfo>,
+    reason: ProvisioningReason,
+}
+
+impl ProvisioningServiceBuilder {
+    fn new() -> Self {
+        Self {
+            reason: ProvisioningReason::Erased,
+            last_connection_attempt: Default::default(),
+            provisioning_info: Default::default(),
+        }
+    }
+    fn with_provisioning_info(mut self, info: ProvisioningInfo) -> Self {
+        let _ = self.provisioning_info.insert(info);
+        self
+    }
+    fn with_reason(mut self, reason: ProvisioningReason) -> Self {
+        self.reason = reason;
+        self
+    }
+    fn with_network_info(mut self, info: NetworkInfo) -> Self {
+        let _ = self.last_connection_attempt.insert(info);
+        self
+    }
```

Comment:
doing this will do a "move out of mutable reference"  when chaining calls. We could wrap it in an inner type like for fs::OpenOptions but seems a bit contrive for this purpose